src/api/syscall.c

```diff
@@ -570,6 +570,13 @@ static void handleRecv(bool_t isBlocking)
         handleFault(NODE_STATE(ksCurThread));
         break;
     }
+
+#ifdef CONFIG_KERNEL_MCS
+    if (isRunnable(NODE_STATE(ksCurThread)) && !ensureSchedulable(NODE_STATE(ksCurThread), true)) {
```

Comment:
this is contrary to the initial design of the kernel where we *always* call rescheduleRequired when we do something that could require a reschedule, rather than checking at the end of every single system call. 

Also, I'm not sure you have enough state here to make this decision - how do you know at this point that ksCurThread is actually going to be scheduled? we could have already switched it to blocked? In which case, are you triggering a reschedule unnecessarily?

Reply:
This may have been an oversimplification of the common case at the end of `receiveIPC` and `receiveSignal`.

Reply:
On review, I think this is fine. If it is blocked, the `isRunnable` will return false. If it is runnable and (i.e. `ThreadState_Running`, `ThreadState_Restate`, or `ThreadState_RunningVM`) then it hasn't been blocked but it may have received a new scheduling context at this point or may have no scheduling context after donating the `ksCurSC` in the first phase (if there was one).

This is also only for `habdleRecv` rather than every syscall, so this is only on receive from an endpoint or notification.

Reply:
if its possible to remove branches then you should

Reply:
This branch is required on all invocations of `handleRecv` as it is always possible that this has occurred after the current thread donated its SC in the first phase. If it has, there is always the potential that it reaches the end of `handleRecv` without an SC, regardless of what kind of IPC object is used for the receive.